### PR TITLE
Support for ctest GPU allocations using auto ctest resource file (trilinos/Trilinos#2422)

### PR DESCRIPTION
**Pull request by bartlettroscoe**
This implements support for using the new CTest test resource file feature in CTest 3.18+ (and CMake 3.17 with a patch installed by [`TriBITS/tribits/devtools_install/install-cmake.py`](https://github.com/TriBITSPub/TriBITS/blob/master/tribits/devtools_install/install-cmake.py)) that spreads work evenly over the provided GPUs.  See motivation in trilinos/Trilinos#2422.  The matching Trilinos PR for this is trilinos/Trilinos#6840.  (The TriBITS commits from that PR branch will be removed.)

### CODE REVIEW

File: test/core/ExamplesUnitTests/CMakeLists.txt [hunk 1]
```diff
@@ -841,6 +841,228 @@ TRIBITS_ADD_ADVANCED_TEST( TribitsHelloWorld_install_config_dummy_proj
...
+  # autogenerate option is enabled!
+
+
+TRIBITS_ADD_ADVANCED_TEST( TribitsHelloWorld_set_default_ctest_resources_file_autogen_off
+  OVERALL_WORKING_DIRECTORY TEST_NAME
+  OVERALL_NUM_MPI_PROCS 1
+  EXCLUDE_IF_NOT_TRUE HAS_CTEST_RESOURCE_FILE_SUPPORT
+
+  TEST_0
+    MESSAGE "Copy in pre-created ctest resource file into the default autogen name and location"
```
Comment (KyleFromKitware) [inline test/core/ExamplesUnitTests/CMakeLists.txt:1033]: There is no need to copy this in - you can just set `CTEST_RESOURCE_SPEC_FILE` to `${CMAKE_CURRENT_SOURCE_DIR}/ctest_resources/other_ctest_resources.json`.

File: test/core/ExamplesUnitTests/CMakeLists.txt [hunk 2]
```diff
@@ -841,6 +841,228 @@ TRIBITS_ADD_ADVANCED_TEST( TribitsHelloWorld_install_config_dummy_proj
...
+      -DTribitsHelloWorld_CUDA_SLOTS_PER_GPU=3   # Will not get used!
+      -DTPL_ENABLE_CUDA=ON
+      -DCTEST_RESOURCE_SPEC_FILE=other_ctest_resources.json
+      ${${PROJECT_NAME}_TRIBITS_DIR}/examples/TribitsHelloWorld
+    PASS_REGULAR_EXPRESSION_ALL
+      "NOTE: The test resource file CTEST_RESOURCE_SPEC_FILE='.*/TriBITS_TribitsHelloWorld_set_other_ctest_resources_file_autogen_on/other_ctest_resources[.]json' will not be auto-generated even through TribitsHelloWorld_AUTOGENERATE_TEST_RESOURCE_FILE=ON because its location does not match the default location '.*/TriBITS_TribitsHelloWorld_set_other_ctest_resources_file_autogen_on/ctest_resources[.]json'[.]  If you want to auto-generate this file, please clear CTEST_RESOURCE_SPEC_FILE and reconfigure or create that file on your own and clear TribitsHelloWorld_AUTOGENERATE_TEST_RESOURCE_FILE[.]"
+      "Configuring done"
+    ALWAYS_FAIL_ON_NONZERO_RETURN
+
+  TEST_2
```
Comment (KyleFromKitware) [inline test/core/ExamplesUnitTests/CMakeLists.txt:969]: Shouldn't this have `ALWAYS_FAIL_ON_NONZERO_RETURN`?

File: test/core/ExamplesUnitTests/CMakeLists.txt [hunk 3]
```diff
@@ -841,6 +841,228 @@ TRIBITS_ADD_ADVANCED_TEST( TribitsHelloWorld_install_config_dummy_proj
...
+      -DTribitsHelloWorld_CUDA_NUM_GPUS=2
+      -DTribitsHelloWorld_CUDA_SLOTS_PER_GPU=3
+      -DTPL_ENABLE_CUDA=ON
+      -DCTEST_RESOURCE_SPEC_FILE=ctest_resources.json
+      ${${PROJECT_NAME}_TRIBITS_DIR}/examples/TribitsHelloWorld
+    PASS_REGULAR_EXPRESSION_ALL
+      "Configuring done"
+    ALWAYS_FAIL_ON_NONZERO_RETURN
+
+  TEST_2
```
Comment (KyleFromKitware) [inline test/core/ExamplesUnitTests/CMakeLists.txt:1014]: Shouldn't this have `WILL_FAIL`?

File: test/core/ExamplesUnitTests/CMakeLists.txt (hunk 2 above)
Comment (bartlettroscoe) [reply test/core/ExamplesUnitTests/CMakeLists.txt:969]: @KyleFromKitware,

No, the default pass/fail criteria is for a zero return code what is the case here.

The options the `ALWAYS_FAIL_ON_NONZERO_RETURN` and `ALWAYS_FAIL_ON_ZERO_RETURN` were added to the `TRIBITS_ADD_ADVANCED_TEST()` function to address the case where you want to determine pass/fail based on a regex **and** you want to ensure the return code is zero or non-zero.   See:

* https://tribits.org/doc/TribitsDevelopersGuide.html#test-case-pass-fail-tribits-add-advanced-test

Basic CMake/CTest does not support an option like.  You have to either check the STDOUT or the return code but you can't check both.  We have found that important defects slip though if you don't check both the STDOUT and the return code.   CTest should really add options like  `ALWAYS_FAIL_ON_NONZERO_RETURN` and `ALWAYS_FAIL_ON_ZERO_RETURN`  to be complete (or you force developers to write their own test runners to be safe).

Does that make sense?

Comment (KyleFromKitware) [reply test/core/ExamplesUnitTests/CMakeLists.txt:969]: Yes, thanks for the explanation.

**Comment by bartlettroscoe** [comment]
@KyleFromKitware, are you finished your review of this PR?  I was wanting to merge this tonight and get this snapshotted to Trilinos in PR https://github.com/trilinos/Trilinos/pull/6840.  If not, we can make changes after the merge to in a new PR.

**Comment by KyleFromKitware** [comment]
> @KyleFromKitware, are you finished your review of this PR? I was wanting to merge this tonight and get this snapshotted to Trilinos in PR [trilinos/Trilinos#6840](https://github.com/trilinos/Trilinos/pull/6840). If not, we can make changes after the merge to in a new PR.

Yes, it looks good to me.

**Pull request merged**